marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mapping/entity-services/entity-validation-lib.sjs

```diff
@@ -0,0 +1,223 @@
+      formattedErrorMessage = {
+        propertyName: propertyName,
+        message: "Required " + errorMessage
+      };
+    }
+  }
+  return formattedErrorMessage;
+}
+
+function validateXmlEntity(newInstance, options, validateEntityValue) {
+  const result = fn.head(xdmp.xqueryEval(
+    'declare variable $newInstance external; xdmp:validate($newInstance, "strict")',
+    {newInstance: newInstance}
+  ));
+
+  if (result != null) {
+    let errorCount = result.xpath("count(/*:error)");
+    if (errorCount > 0) {
+      if ("accept" == validateEntityValue) {
+        if (options.headers == null) {
+          options.headers = {};
+        }
+        if (options.headers.datahub == null) {
+          options.headers.datahub = {};
+        }
+        options.headers.datahub.validationErrors = [];
+
+        for (error of result.xpath("/*:error")) {
+          let validationError = {
+            error : {
+              code: fn.string(error.xpath("./*:code/text()")),
+              name: fn.string(error.xpath("./*:name/text()")),
+              message: fn.string(error.xpath("./*:message/text()")),
+              formatString: fn.string(error.xpath("./*:format-string/text()"))
+            }
+          };
+          addFormattedMessagesForXml(validationError);
+          options.headers.datahub.validationErrors.push(validationError);
+        }
+      } else if ("reject" == validateEntityValue) {
+        throw Error(result);
+      }
+    }
+  }
+}
+
+function addFormattedMessagesForXml(validationError) {
+  if (validationError.error.code == "XDMP-VALIDATEMISSINGELT") {
+    buildFormattedMessagesForMissingElementsError(validationError);
+  }
+  else if (validationError.error.code == "XDMP-VALIDATEUNEXPECTED") {
+    buildFormattedMessagesForUnexpectedNodeError(validationError);
+  }
+}
+
+/**
+ * Example of what the formatString is expected to contain: "Missing required elements Expected (LastName,Email?)"
+ *
+ * @param validationError
+ */
+function buildFormattedMessagesForMissingElementsError(validationError) {
+  let formatString = validationError.error.formatString;
+  let indicator = "Expected (";
+  let pos = formatString.indexOf(indicator);
+  if (pos > -1) {
+    let str = formatString.substring(pos + indicator.length);
+    pos = str.indexOf(")");
+    if (pos > -1) {
+      str = str.substring(0, pos);
+      let names = str.split(",");
+      let notRequiredPos = names.findIndex(val => val.endsWith("?"));
+      if (notRequiredPos > -1) {
+        validationError.error.formattedMessages = buildFormattedMessagesForMissingXmlPropertyNames(names.slice(0, notRequiredPos))
+      }
+    }
+  }
+}
+
+/**
+ * Example of what the formatString is expected to contain: "Invalid node: Found LastName but expected (FirstName,LastName,Email?)"
+ *
+ * @param validationError
+ */
+function buildFormattedMessagesForUnexpectedNodeError(validationError) {
+  let formatString = validationError.error.formatString;
```

Comment:
Null checks please

Reply:
validationError.error is guaranteed to exist here

Reply:
No it's not. We do not do raw dot notation on objects without checks, these lead to uncaught runtime exceptions.